source/extensions/http/header_validators/envoy_default/http1_header_validator.cc

```diff
@@ -0,0 +1,416 @@
+#include "source/extensions/http/header_validators/envoy_default/http1_header_validator.h"
+
+#include "source/extensions/http/header_validators/envoy_default/character_tables.h"
+
+#include "absl/container/node_hash_set.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Http {
+namespace HeaderValidators {
+namespace EnvoyDefault {
+
+using ::envoy::extensions::http::header_validators::envoy_default::v3::HeaderValidatorConfig;
+using ::Envoy::Http::HeaderString;
+using ::Envoy::Http::LowerCaseString;
+using ::Envoy::Http::Protocol;
+using ::Envoy::Http::RequestHeaderMap;
+using HeaderValidatorFunction = ::Envoy::Http::HeaderValidator::HeaderEntryValidationResult (
+    Http1HeaderValidator::*)(const HeaderString&);
+
+struct Http1ResponseCodeDetailValues {
+  const std::string InvalidTransferEncoding = "uhv.http1.invalid_transfer_encoding";
+  const std::string TransferEncodingNotAllowed = "uhv.http1.transfer_encoding_not_allowed";
+  const std::string ContentLengthNotAllowed = "uhv.http1.content_length_not_allowed";
+  const std::string ChunkedContentLength = "uhv.http1.content_length_and_chunked_not_allowed";
+};
+
+using Http1ResponseCodeDetail = ConstSingleton<Http1ResponseCodeDetailValues>;
+
+/*
+ * Header validation implementation for the Http/1 codec. This class follows guidance from
+ * several RFCS:
+ *
+ * RFC 3986 <https://datatracker.ietf.org/doc/html/rfc3986> URI Generic Syntax
+ * RFC 9110 <https://www.rfc-editor.org/rfc/rfc9110.html> HTTP Semantics
+ * RFC 9112 <https://www.rfc-editor.org/rfc/rfc9112.html> HTTP/1.1
+ *
+ */
+Http1HeaderValidator::Http1HeaderValidator(const HeaderValidatorConfig& config, Protocol protocol,
+                                           StreamInfo::StreamInfo& stream_info)
+    : HeaderValidator(config, protocol, stream_info) {}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http1HeaderValidator::validateRequestHeaderEntry(const HeaderString& key,
+                                                 const HeaderString& value) {
+  static const absl::node_hash_map<absl::string_view, HeaderValidatorFunction> kHeaderValidatorMap{
+      {":method", &Http1HeaderValidator::validateMethodHeader},
```

Comment:
Nit: I would add a comment that pseudo headers are synthesized by the codec from the request line before submitting header map for validation, since people might get confused.

Reply:
ping!  I in fact had this same confusion and went looking for a comment :-P